React Native - SVG icons for Inserter

mzorz opened the pull request:
## Description
This PR adds mobile SVG compatibility in React Native for SVG icons of already ported blocks:
- Code
- Paragraph
- More
- Heading

## How has this been tested?
Using [this PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/129) in the Gutenberg mobile repo

## Screenshots
<img width="262" alt="screen shot 2018-08-23 at 21 48 42" src="https://user-images.githubusercontent.com/6597771/44560038-fae7f680-a723-11e8-910b-b9bdd24b347c.png">

## Types of changes
This PR 
- adds support for the SVG `<g>` tag in the for of `G` as needed by our wrapper library
- changes the `icon:` prop declaration for the aforementioned blocks to use the wrapper `<SVG` tag as follows:

What was til now declared as:
```
	icon: <svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path fill="none" d="M0 0h24v24H0V0z" /><g><Path d="M2 9v2h19V9H2zm0 6h5v-2H2v2zm7 0h5v-2H9v2zm7 0h5v-2h-5v2z" /></g></svg>,
```

is now declared as:
```
	icon: <SVG viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><Path fill="none" d="M0 0h24v24H0V0z" /><G><Path d="M2 9v2h19V9H2zm0 6h5v-2H2v2zm7 0h5v-2H9v2zm7 0h5v-2h-5v2z" /></G></SVG>,
```

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

mzorz committed:
exports SVG <G> tag and adds mobile compatible SVG icons for code, heading, more and paragraph blocks

mzorz committed:
adds simple mechanism to obtain viewport from viewBox so height and width do not have to be indicated explicitely

gziolo commented:
There are some linting issues which prevents this PR from being merged. You can see them on Travis: https://travis-ci.org/WordPress/gutenberg/jobs/419924102

```
[0] [0] /home/travis/build/WordPress/gutenberg/packages/block-library/src/code/index.js
[0] [0]   21:1  error  Trailing spaces not allowed  no-trailing-spaces
[0] [0] 
[0] [0] /home/travis/build/WordPress/gutenberg/packages/block-library/src/heading/index.js
[0] [0]   24:3  error  Trailing spaces not allowed  no-trailing-spaces
[0] [0] 
[0] [0] /home/travis/build/WordPress/gutenberg/packages/components/src/primitives/svg/index.native.js
[0] [0]   10:5   error  There must be a space inside this paren                   space-in-parens
[0] [0]   10:18  error  Expected '!==' and instead saw '!='                       eqeqeq
[0] [0]   10:47  error  Expected '!==' and instead saw '!='                       eqeqeq
[0] [0]   10:59  error  There must be a space inside this paren                   space-in-parens
[0] [0]   16:9   error  Unnecessary 'else' after 'return'                         no-else-return
[0] [0]   19:7   error  'viewBoxCoords' is never reassigned. Use 'const' instead  prefer-const
[0] [0]   19:42  error  There must be a space inside this paren                   space-in-parens
[0] [0]   19:46  error  There must be a space inside this paren                   space-in-parens
[0] [0]   21:30  error  A space is required after '['                             computed-property-spacing
[0] [0]   21:32  error  A space is required before ']'                            computed-property-spacing
[0] [0]   21:58  error  A space is required after '['                             computed-property-spacing
[0] [0]   21:60  error  A space is required before ']'                            computed-property-spacing
```

gziolo commented:
@mtias @jasmussen - out of curiosity, how are those new SVG icons generated? I'm asking because we need to take into account the needs of React Native environment moving forward where we will have to use imported primitive components which allow to branch code between web and mobile.

gziolo commented on packages/components/src/primitives/svg/index.native.js:
```diff
@@ -1,13 +1,26 @@
+				{ props.children }
+			</Svg>
+		);
+	} else {
```
Svg implementation handles `viewBox` out of the box:
https://github.com/react-native-community/react-native-svg/blob/master/elements/Svg.js#L35

mzorz replied:
That's right `viewBox` is handled, but the component will fail to render if `height`  and `width` are not set - what this  (most probably hack-ish) piece of code does is fabricate the height and width out of the information passed in `viewBox`  which seemed to exist in all cases for all blocks in code (I relied on that to process unless `width` and `height` are explicitely set through props). For this I separated this specific change in commit c75a0db13fe675b80cfd869aa926256b0a7c5553  wdyt?

jasmussen commented:
>  out of curiosity, how are those new SVG icons generated? I'm asking because we need to take into account the needs of React Native environment moving forward where we will have to use imported primitive components which allow to branch code between web and mobile.

That's the thing. We can't know. Because plugins add these, so they could be using horribly mangled SVGs.

As is also discussed in https://github.com/WordPress/gutenberg/issues/9269#issuecomment-415397686, this also has accessibility implications. 

Do we have any way to build a tool to parse the SVG source files and serve a sanitized version on the fly?

SergioEstevao commented on packages/components/src/primitives/svg/index.native.js:
```diff
@@ -1,13 +1,26 @@
+		// take viewport system to match the viewBox definition
+		// i.e. viewBox="0 0 24 24"
+		let viewBoxCoords = props.viewBox.split(' ');
+		return (
```
That was the problem, RN complained about className being set. if omit takes it away it could work :).

gziolo replied:
I think this is going to be a recurring issue, it would be nice to find a way to skip those `className` properties or convert them to `style`  which would work out of the box.

mzorz replied:
tested it a bit and it seems to be working 👍 
addressed in 6d5b37db0

gziolo commented:
> That's the thing. We can't know. Because plugins add these, so they could be using horribly mangled SVGs.

I wouldn't worry about it too much. We didn't even start discussing what needs to happen to enable plugins on mobile. I can imagine this is going to be opt-in because you will have to operate on a much more constrained environment. You won't be able to use DOM element as they are. SVG is just and example.

> Do we have any way to build a tool to parse the SVG source files and serve a sanitized version on the fly?

I don't think there is one, however it isn't that complex. In the majority of cases, the only thing you need to change is to capitalize tag name. SVG is the only exception because it needs to follow our coding guidelines. Please also note that you can't fully copy `<svg />` as is today, you always need to convert `class` to `className` to make it work with React.

SergioEstevao committed:
Update use of content size change event.

mzorz committed:
fixed lint issues

mzorz committed:
removed unneeded aliases

mzorz commented on packages/components/src/primitives/svg/index.native.js:
```diff
@@ -1,13 +1,26 @@
 } from 'react-native-svg';
 
 export const Path = PATH;
+export const G = GROUP;
```
addressed in bf65beb58

mzorz committed:
using omit from loadsh to exclude className from props when passing down to Svg component

mzorz committed:
moved imports to their corresponding section

mzorz commented on packages/block-library/src/code/index.js:
```diff
@@ -8,6 +8,7 @@ import { createBlock } from '@wordpress/blocks';
  * Internal dependencies
  */
 import edit from './edit';
+import { Path, SVG } from '@wordpress/components';
```
Thanks! addressed in ed725d672

mzorz commented on packages/block-library/src/heading/index.js:
```diff
@@ -20,7 +20,8 @@ import { RichText } from '@wordpress/editor';
  */
 import edit from './edit';
-
+import { Path, SVG } from '@wordpress/components';
```
addressed in ed725d672

mzorz commented on packages/block-library/src/more/index.js:
```diff
@@ -14,6 +14,7 @@ import { createBlock } from '@wordpress/blocks';
  * Internal dependencies
  */
 import edit from './edit';
+import { Path, G, SVG } from '@wordpress/components';
```
same as above

mzorz commented on packages/block-library/src/paragraph/index.js:
```diff
@@ -25,6 +25,7 @@ import {
  * Internal dependencies
  */
 import edit from './edit';
+import { Path, SVG } from '@wordpress/components';
```
:) same as above, addressed in ed725d672

daniloercoli committed:
- Force update the RichText component when a toolbar button is pressed
- Fix an issue where eventCount was not correctly propogated to AztecText resulting in a re-rendering of the component from scratch

mzorz committed:
stripping style if not an object in svg native implementation

mzorz commented:
Added stripping `style` for `<svg>` native implementation as discussed elsewhere, commit has comment explaining 516f66120.

Ready for another round @gziolo 🙇

mzorz committed:
Merge branch 'master' into rnmobile/inserter-svgicons-take1

mzorz committed:
Merge commit 'cb7c413' into HEAD

mzorz committed:
Merge branch 'fix/icon-svg-gb-hash-reference' into rnmobile/inserter-svgicons-take1

mzorz committed:
added empty native implementation of viewport

mzorz committed:
fixed lint issues

mzorz committed:
fixed merge conflict

mzorz committed:
Merge branch 'master' into rnmobile/inserter-svgicons-take1

mzorz committed:
Merge commit '1b5653a' into rnmobile/inserter-svgicons-take1-merge

mzorz commented on packages/components/src/primitives/svg/index.native.js:
```diff
@@ -1,13 +1,33 @@
+};
+
 export const SVG = ( props ) => {
+	// for some reason if a `className` prop is passed it gets converted to `style` here.
```
citing the comment: `Given it carries a string (as it was originally className) but an object is expected for "style"`... it chokes :)

mzorz replied:
Also thanks for pointing that out (about the PR where this change is implemented). We discussed that elsewhere and forgot to add this here. It might be good to change the comment to add this information in particular, adjusting the _"for some reason..." _ part.

mzorz replied:
Comment modified in 522785990

mzorz committed:
Modified comment to better reflect causes of className being conveted to style for RN

hypest commented on packages/components/src/primitives/svg/index.native.js:
```diff
@@ -1,13 +1,33 @@
+	// We're using the react-native-classname-to-style plugin, so when a `className` prop is passed it gets converted to `style` here.
+	// Given it carries a string (as it was originally className) but an object is expected for `style`,
+	// we need to check whether `style` exists and is a string, and strip in that case.
+	const safeProps = ( typeof props.style === 'string' || props.style instanceof String ) ? { ...omit( props, [ 'style', 'className' ] ) } : { ...omit( props, [ 'className' ] ) };
```
Feels super weird to have to manually strip the style definition so, if you don't mind @mzorz , let's dig a little deeper in this one.

We already had a chat over Slack with what probably needs to be done here (properly load the styles for RN) so, let's wait a bit to finish that effort, is that OK?

mzorz replied:
As discussed with @hypest , made the `style` convert from string to object properly (if it applies) in 2d474511dd100604421564ea3cfb745c300f7cc7 and b678c6d2abba31a627dd5eb9d307cb36c49aede1. 

To test, I added the following:
```
.dashicons-editor-bold {
	transform: rotate(20deg);
}

```
to `packages/components/src/dashicon/style.scss` 

and the styles get passed to the svg component, this is how the bold icon gets rendered:

<img width="307" alt="screen shot 2018-09-06 at 13 21 46" src="https://user-images.githubusercontent.com/6597771/45171725-860fc480-b1d9-11e8-9fac-1c0bba64c3f3.png">

See the **B** icon slightly leaning to the right? 😄

gziolo replied:
Out of curiosity, why does it happen that style becomes a string?
On the web you have to pass an object as style, I think this is also the case in React Native. So this is very surprising. I would love to understand why this is something we need to handle in the first place.

mzorz replied:
As far as I understand, `className` is [not supported in React Native](https://github.com/facebook/react-native/issues/7238), so we need to define that as a prop on each object we'd want to use that in.
Because of this and being that `className` is being used in different places in Gutenberg, we're using a [Babel plugin](https://github.com/kristerkari/babel-plugin-react-native-classname-to-style) that transforms `className` to `style` in RN so to make it transparent to Gutenberg code.

AFAIU the plugin just keeps the passed object as it was without making any other conversion, so if the original object was a string, it will be a string after going through it as well.

So knowing this, it looks like the problem arises then because Dashicon `gutenberg/packages/components/src/dashicon/index.js` declares `className={ iconClass }` where `iconClass` is a string, a defined here: https://github.com/WordPress/gutenberg/blob/master/packages/components/src/dashicon/index.js#L899

While `className` can be a string (as it originally is in Dashicon code), `style` should be an object instead, but we're not transforming it (just keeping the same type).

Does that make sense? Should we do something differently? cc @hypest as well to see if there's something else to add / do differently in this regard.

mzorz committed:
building an object out of the passed style/className concatenated string on svgnative implementation

gziolo commented:
This needs to be synced with the web implementation. We discussed it a few times with @jasmussen and there is no clear way to move forward. It would be great if we could figure out a solution where we would be able to use SVG as is generated and convert it to React element on the fly, for web it would be possible with this Babel plugin:
https://github.com/airbnb/babel-plugin-inline-react-svg/blob/master/README.md. It looks like there is a bit different solution for React Native: https://github.com/vault-development/react-native-svg-uri, but maybe we could find something that would work on both platforms.

hypest commented:
@gziolo I wonder, can we perhaps abstract away the value of the `icon` block field into a function and then nativize that function? On the web case that function would return a string (dashicon name) or the lowercase `<svg` React tree, while on the mobile case it will use the `<Svg` variant. Can't that be a path forwart for the time being?

mzorz committed:
fix array level indirection when looking into styles object

gziolo commented:
> I'm curious, what's the reasoning behind inlining SVGs? Could we keep them as a file and use require to load them? Then each platform could use a different loader for them, right?

It's more or less the same, not going into detail, it would work on a higher level as you described.

jasmussen commented:
> I'm curious, what's the reasoning behind inlining SVGs? Could we keep them as a file and use require to load them? Then each platform could use a different loader for them, right?

In the _rendered code_, we need the SVGs to be inline so we can colorize them using `fill` CSS properties. But I'm sort of assuming that subtext of your note implies this would still be the case if we loaded SVG content from files.

jasmussen commented:
One of the challenges we are also facing (see #9565) is that we need a few attributes added to the SVG, which we can't assume are always there.

gziolo commented:
I'm playing with an alternative approach in #9685. The idea is as follows:

> To make it work with React Native platform, I want to overload `createElement` to use `react-native-svg` elements instead of HTML types:
> - `svg` -> `Svg` (from `react-native-svg`)
> - `path` -> `Path`
> - `g` -> `G`
> - and so on
> 
> This would be applied only to the RN version of `react` implementation.

gziolo committed:
Merge remote-tracking branch 'origin/master' into rnmobile/inserter-svgicons-take1

hypest commented:
From https://github.com/WordPress/gutenberg/pull/9685#issuecomment-425344400:
> Okey, let's merge this one and rebase #9294 against master.

@mzorz , let's try updating from `master` here and see where this goes, wdyt?

mzorz committed:
fixed merge conflicts

mzorz commented:
> @mzorz , let's try updating from master here and see where this goes, wdyt?

Just finished updating this branch to `master` - also updated the related Gutenberg mobile PR at https://github.com/wordpress-mobile/gutenberg-mobile/pull/129
@hypest @gziolo

mzorz committed:
enforcing usage of SVG and Path primitives in image icon field for React Native compatibility

hypest reviewed (approved):
LGTM! Feel free to merge as soon as Travis gives the green light too!
